File: 20-valid-parentheses/valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        std::stack <char> stack;

        for (auto chr : s) {
            if (chr == '(') {
                stack.push('(');
            } else if (chr == '{') {
                stack.push('{');
            } else if (chr == '[') {
                stack.push('[');
            }
            if (chr == ')' ) {
               if (stack.size() != 0 && stack.top()=='(') {stack.pop();}
               else {return false;}
            } else if (chr == '}'  ) {
                if (stack.size() != 0 && stack.top()=='{') {stack.pop();}
                else {return false;}
            } else if (chr == ']'  ) {
                if (stack.size() != 0 && stack.top() == '[' ){stack.pop();}
                else {return false;}
            }
        }
        if (stack.size() == 0) return true;
        return false;
    }
};

File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        
        stack = []
        
        def isTopValid(par : str, top : str) -> bool:
            if top == "(" and par == ")":
                return True
            if top == "{" and par == "}":
                return True                
            if top == "[" and par == "]":
                return True         
            return False
            

        for letter in s:
            if letter == "(" or letter == "{" or letter == "[":
                stack.append(letter)
                continue
            if letter == ")" or letter == "}" or letter == "]":
                if len(stack) < 1:
                    return False
                if isTopValid(letter, stack[-1]):
                    stack.pop()
                else:
                    return False
        
        return len(stack) == 0
            
            



                


